tests/test_github_repo_request_local.py

```diff
@@ -0,0 +1,189 @@
+import pytest
+from src.github_repo_request_local import get_base_repo_url, filter_incomplete_urls
+import pandas as pd
+
+
+def test_get_base_repo_url():
+    # Empty or null input
+    assert get_base_repo_url("") is None
+    assert get_base_repo_url(None) is None
+
+    # Valid URLs
+    assert (
+        get_base_repo_url("https://github.com/getdnsapi/stubby")
+        == "https://github.com/getdnsapi/stubby"
+    )
+
+    assert get_base_repo_url("https://github.com/namecoin") is None
+
+    assert (
+        get_base_repo_url("https://github.com/osresearch/heads/issues/540")
+        == "https://github.com/osresearch/heads"
+    )
+
+    # URLs from other code hosting platforms
+    assert (
+        get_base_repo_url("https://git.sr.ht/~dvn/boxen")
+        == "https://git.sr.ht/~dvn/boxen"
+    )
+
+    assert (
+        get_base_repo_url(
+            "https://redmine.replicant.us/projects/replicant"
+            "/wiki/Tasks_funding#Graphics-acceleration"
+        )
+        == "https://redmine.replicant.us/projects/replicant"
+    )
+
+    # Checking the `repourls` which ends with `.git`
+    assert (
+        get_base_repo_url("https://github.com/stalwartlabs/mail-server.git")
+        == "https://github.com/stalwartlabs/mail-server"
+    )
+
+    assert (
+        get_base_repo_url("https://github.com/tdf/odftoolkit.git")
+        == "https://github.com/tdf/odftoolkit"
+    )
+
+    # Invalid URL formats
+    assert get_base_repo_url("just-a-string") is None
+    assert get_base_repo_url("http:///a-bad-url.com") is None
+
+    # URLs with query parameters or fragments
+    assert (
+        get_base_repo_url("https://github.com/getdnsapi/stubby.git#readme")
+        == "https://github.com/getdnsapi/stubby"
+    )
+
+    assert (
+        get_base_repo_url("https://github.com/getdnsapi/stubby.git?branch=main")
+        == "https://github.com/getdnsapi/stubby"
+    )
+
+
+@pytest.mark.parametrize(
+    "data, expected_length, expected_urls",
+    [
+        # Test with an empty DataFrame
+        ({"repourl": []}, 0, []),
+        # DataFrame without the 'repourl' column
+        ({"other_column": ["data1", "data2"]}, 2, []),
+        # DataFrame with various types of URLs including None and empty strings
+        (
+            {
+                "repourl": [
+                    "https://github.com/user/repo",
```

Comment:
As a detail, GitHub uses the term `owner` rather than `user` and I think `owner` is more reflective of/for many of the repos we're processing as they're owned by an organisation rather than the property of an individual person. So let's use `owner` please unless there are good reasons to prefer `user`.